### PR TITLE
[FEATURE] Search Schemes

### PR DESCRIPTION
**Pull request by cpockrandt**
~Do not review or merge yet!~ :)

### CODE REVIEW

**Comment by marehr** [comment]
Optimum?

**Comment by cpockrandt** [comment]
@h-2 Yes, optimum.

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    std::vector<uint8_t> l;
+    std::vector<uint8_t> u;
+
+    uint8_t blocks() const noexcept
+    {
+        return pi.size();
+    }
+};
+
+using search_scheme_dyn = std::vector<search_dyn>;
```
Comment (h-2) [inline include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: so search_scheme_dyn **is** a range of searches, but `optimum_search_scheme` **contains** a range of searches (as ::value)?

That's a weird design :thinking: 

I would instead make the optimum_search_scheme a variable template:
```cpp
template <uint8_t min_errors, uint8_t max_errors>
void optimum_search_scheme;

template <>
inline search_scheme<1, 1> constexpr optimum_search_scheme<0,0> = {{ .... }};

...
```
This has the nice side-effect that if two optimum_search_schemes have the same "value type" they are also the same type and you don't have double-code instantiation.

Comment (h-2) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: But then also change the naming so that "search_scheme" doesn't refer to both types and values.

Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: I agree with everything but you cannot declare it of type `void`? :thinking:

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+        uint64_t lb2 = lb - !go_right;
+        uint64_t rb2 = rb + go_right;
+
+        do
+        {
+            bool const delta = it.last_char() != query[(go_right ? rb : lb) - 1];
+
+            // NOTE: move that outside the if / do-while struct
+            // NOTE: check that as well before doing non-edit-distance steps inside the loop
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: I will add a "pre-merge" note for those TODOs

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 3]
```diff
@@ -77,7 +78,7 @@ inline auto search_single(index_t const & index, auto & query, auto const & cfg)
     std::vector<typename index_t::iterator_type> internal_hits;
     auto internal_delegate = [&internal_hits, &max_error](auto const & it)
     {
-        internal_hits.push_back(it);
+        internal_hits.push_back(it); // TODO: report max_error-error_left
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search.hpp]: We should discuss this once more since we have different opinions on what should be returned.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+//!\brief Computes a (not optimal) search scheme.
+inline auto compute_search_scheme(uint8_t const min_error, uint8_t const max_error)
+{
+    // TODO: Simple backtracking. Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
+
+//!\brief Returns for each search the cumulative length of blocks in the order of blocks in each search.
+inline auto search_scheme_block_info(auto const & search_scheme, uint64_t const query_length)
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: We could but
* search schemes only occur in detail namespace
* the compiler will build a lot of code paths with search schemes (0-4 errors, different combinations of configuration objects, etc.)

IMO this would only increase the compilation time and doesn't pay off just to require the parameter `search_scheme` to fulfill the `search_scheme_concept`.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+template <bool abort_on_hit>
+inline bool search_scheme_single_search_exact(auto it, auto & query, uint64_t const lb, uint64_t const rb,
+                                              uint8_t const errors_spent, uint8_t const block_id, bool const go_right,
+                                              auto const & search, auto const & blocklength,
+                                              search_params const error_left, auto && delegate)
+{
+    uint8_t const block_id2 = std::min(block_id + 1, static_cast<uint8_t>(search.blocks()) - 1);
+    bool const go_right2 = (block_id < search.blocks() - 1) && search.pi[block_id + 1] > search.pi[block_id];
+    // NOTE: from sven:
+    // bool const goToRight2 = (blockIndex < s.pi.size() - 1) ? s.pi[blockIndex + 1] > s.pi[blockIndex] : s.pi[blockIndex] > s.pi[blockIndex - 1];
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: This is an open pre-merge TODO

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+{
+    // TODO: Simple backtracking. Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
+
+//!\brief Returns for each search the cumulative length of blocks in the order of blocks in each search.
+inline auto search_scheme_block_info(auto const & search_scheme, uint64_t const query_length)
+{
+    // TODO: use std::array for optimum_search_scheme, std::vector otherwise
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: Any idea on how to do it in an elegant way (i.e. without overloading functions, code duplication, etc.)?

It seems that there isn't a way with `std::conditional` since I need to retrieve the size of the array with `std::tuple_size` which is not defined for `std::vector`. Contrary to `if constexpr` the code has to be correct for the false-case as well

**Comment by cpockrandt** [comment]
Can be reviewed now :)

**Comment by cpockrandt** [comment]
!rebuild

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+
+    // Either store information in an array (for search schemes known at compile time) or in a vector otherwise.
+    using result_type = typename std::conditional_t<std::is_same_v<search_scheme_t, search_scheme_dyn_type>,
+                                                    std::vector<std::tuple<blocks_length_type, uint64_t>>,
+                                                    std::array<std::tuple<blocks_length_type, uint64_t>,
+                                                               transformation_trait_or_t<std::tuple_size<search_scheme_t>,
+                                                                                         std::false_type>::value>>;
+
+    result_type result;
+    if constexpr (std::is_same_v<result_type, std::vector<std::tuple<blocks_length_type, uint64_t>>>)
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
    if constexpr (std::Same<result_type, std::vector<std::tuple<blocks_length_type, uint64_t>>>)
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+        result.resize(search_scheme.size());
+
+    uint8_t  const blocks      {search_scheme[0].blocks()};
+    uint64_t const block_length{query_length / blocks};
+    uint8_t  const rest        {query_length - blocks * block_length};
+
+    blocks_length_type blocks_length;
+    // set all blocks_length values to block_length
+    // resp. block_length + 1 for the first `rest = block_length % blocks` values
+    if constexpr (std::is_same_v<blocks_length_type, std::vector<uint64_t>>)
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
    if constexpr (std::Same<blocks_length_type, std::vector<uint64_t>>)
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+
+    for (uint8_t search_id = 0; search_id < search_scheme.size(); ++search_id)
+    {
+        auto const & search = search_scheme[search_id];
+
+        auto & [search_blocks_length, start_pos] = result[search_id];
+
+        // compute cumulative blocks_length and starting position
+        start_pos = 0;
+        if constexpr (std::is_same_v<blocks_length_type, std::vector<uint64_t>>)
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
        if constexpr (std::Same<blocks_length_type, std::vector<uint64_t>>)
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+
+//!\cond
+// forward declaration
+template <bool abort_on_hit>
+inline bool search_ss(auto it, auto & query, uint64_t const lb, uint64_t const rb, uint8_t const errors_spent,
+                      uint8_t const block_id, bool const go_right, auto const & search, auto const & blocks_length,
+                      search_param const error_left, auto && delegate);
+//!\endcond
+
+/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
/*!\brief Searches a query sequence in a bidirectional index using a single search of a search scheme.
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+ *        Sub-function for searching the remaining part of the current block without any errors.
+ * \tparam abort_on_hit If the flag is set, the search aborts on the first hit.
+ * \param[in] it Iterator of atring index built on the text that will be searched.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
 * \param[in] it Iterator of a string index built on the text that will be searched.
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+ * \param[in] query Query sequence to be searched with the iterator.
+ * \param[in] lb Left bound of the infix of `query` already searched (exlusive).
+ * \param[in] rb Right bound of the infix of `query` already searched (exlusive).
+ * \param[in] errors_spent Number of errors spent while searching the infix of `query`.
+ * \param[in] block_id Id of the block that infix is extended in next.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
 * \param[in] block_id Id of the block that infix is extended to next.
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+inline bool search_ss(auto it, auto & query, uint64_t const lb, uint64_t const rb, uint8_t const errors_spent,
+                      uint8_t const block_id, bool const go_right, auto const & search, auto const & blocks_length,
+                      search_param const error_left, auto && delegate);
+//!\endcond
+
+/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
+ *        Sub-function for searching the remaining part of the current block without any errors.
+ * \tparam abort_on_hit If the flag is set, the search aborts on the first hit.
+ * \param[in] it Iterator of atring index built on the text that will be searched.
+ * \param[in] query Query sequence to be searched with the iterator.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```suggestion
 * \param[in] query Query sequence to be searched.
```

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+        if (search_ss<abort_on_hit>(it, query, infix_lb, rb, errors_spent, block_id2, go_right2, search, blocks_length,
+                                    error_left, delegate) && abort_on_hit)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]:
```
/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
              Sub-function for deletions at the end of a block.
* \copydetails search_ss_exact
```
Should work (maybe you need the `search_ss_exact` with the namespace. Otherwise apply review from `search_ss_exact`.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+                                                 error_left2, delegate) && abort_on_hit)
+            {
+                return true;
+            }
+        } while ((go_right && it.cycle_back()) || (!go_right && it.cycle_front()));
+    }
+    return false;
+}
+
+/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
```
Comment (eseiler) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: Apply review from `ss_search_exact`
Same goes for the other docs

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 16]
```diff
@@ -34,46 +34,51 @@
...
+        // NOTE: Casting doubles rounds towards zero (i.e. floor for positive numbers). Thus, given a rate of
+        // 10% and a read length of 101 the max number of errors is correctly casted from 10.1 errors to 10
         std::tie(total, subs, ins, del) = std::apply([& query](auto && ... args)
             {
                 return std::tuple{(args * query.size())...};
             }, get<search_cfg::id::max_error_rate>(cfg));
     }
 
-    // TODO: if total not set: max_error.total = max_error.deletion + max_error.substitution + max_error.insertion;
+    // TODO: if total not set: max_error.total = max_error.deletion + max_error.substitution
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search.hpp]: No, all remaining TODOs will be resolved in other pull requests

File: include/seqan3/search/algorithm/search.hpp [hunk 17]
```diff
@@ -46,12 +46,27 @@
 namespace seqan3
 {
 
-//!\brief \todo Document!
+/*!\brief Search a query or a range of queries in an index.
+ * \param[in] index String index to be searched.
+ * \param[in] queries A single query or a (forward) range of queries. A query must model
+                      std::ranges::RandomAccessRange.
+ * \param[in] cfg A configuration object specifying the search parameters (e.g. number of errors, error types,
+ *                output format, etc.).
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: document `\tparam` and return type/value.

File: include/seqan3/search/algorithm/search.hpp [hunk 18]
```diff
@@ -46,12 +46,27 @@
...
+ *
+ * ### Complexity
+ *
+ * Exponential in the numbers of errors.
+ *
+ * ### Exceptions
+ *
+ * No-throw guarantee.
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: see above comments regarding complexity and exceptions. Also make statements about thread-safety

File: include/seqan3/search/algorithm/search.hpp [hunk 19]
```diff
@@ -46,12 +46,27 @@
...
+ */
 template <typename index_t, typename queries_t, typename config_t>
 //!\cond
     requires
         (std::ranges::RandomAccessRange<queries_t> ||
-            (std::ranges::ForwardRange<queries_t> && std::ranges::RandomAccessRange<value_type_t<queries_t>>)) &&
+            (std::ranges::ForwardRange<queries_t> &&
+             std::ranges::RandomAccessRange<value_type_t<queries_t>>)) &&
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: are you sure this is sufficient? Do you not require size for example?

File: include/seqan3/search/algorithm/search.hpp [hunk 20]
```diff
@@ -46,12 +46,27 @@
...
         detail::is_algorithm_configuration_v<remove_cvref_t<config_t>>
 //!\endcond
 inline auto search(index_t const & index, queries_t && queries, config_t const & cfg)
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: what about constraints for the index?

File: include/seqan3/search/algorithm/search.hpp [hunk 21]
```diff
@@ -60,24 +75,25 @@ inline auto search(index_t const & index, queries_t && queries, config_t const &
...
         if (subs > total)
-            throw std::invalid_argument("The substitution error threshold is higher than the total error threshold.");
+            throw std::invalid_argument("Substitution error threshold is higher than the total error "
+                                        "threshold.");
         if (ins > total)
-            throw std::invalid_argument("The insertion error threshold is higher than the total error threshold.");
+            throw std::invalid_argument("Insertion error threshold is higher than the total error threshold.");
         if (del > total)
-            throw std::invalid_argument("The deletion error threshold is higher than the total error threshold.");
+            throw std::invalid_argument("Deletion error threshold is higher than the total error threshold.");
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: why did you remove the articles?

File: include/seqan3/search/algorithm/search.hpp [hunk 22]
```diff
@@ -95,18 +111,31 @@ inline auto search(index_t const & index, queries_t && queries, config_t const &
...
+ * ### Exceptions
+ *
+ * No-throw guarantee.
+ */
 template <typename index_t, typename queries_t>
 //!\cond
     requires std::ranges::RandomAccessRange<queries_t> ||
              (std::ranges::ForwardRange<queries_t> && std::ranges::RandomAccessRange<value_type_t<queries_t>>)
 //!\endcond
 inline auto search(index_t const & index, queries_t && queries)
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: see previous comments

File: test/unit/search/helper.hpp [hunk 23]
```diff
@@ -33,8 +33,7 @@
 // ============================================================================
 
 /*!\file
- * \brief Provides a simple function that sorts a vector and returns it to allow the use of EXCEPT_EQ() for
- *        vectors of arbitrary order while keeping the detailed output for test failures.
+ * \brief Provides helper functions for testing the search modules.
  * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
  */
```
Comment (h-2) [inline test/unit/search/helper.hpp]: tests don't get doxygen

File: test/unit/search/helper.hpp [hunk 24]
```diff
@@ -43,14 +42,16 @@
...
 {
 
 template <typename T>
 std::vector<T> sort(std::vector<T> v)
 {
     std::sort(v.begin(), v.end());
-    v.erase(std::unique(v.begin(), v.end()), v.end()); // TODO: remove once duplicates are filtered
+    v.erase(std::unique(v.begin(), v.end()), v.end());
     return v;
 }
```
Comment (h-2) [inline test/unit/search/helper.hpp]: what does this do? :thinking: 

Remove all elements that don't appear multiple times?

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+    detail::search_param error_left{total, substitution, 0, 0};
+
+    // Find all hits using search schemes.
+    detail::search_ss<false>(it, query, start_pos, start_pos + 1, 0, 0, true, search, blocks_length, error_left,
+                             delegate_ss);
+    // Find all hits using trivial backtracking.
+    detail::search_trivial<false>(it, query, 0, error_left, delegate_trivial);
+
+    // Eliminate hits that we are not interested in (based on the search and chosen error distribution)
+    hits_ss.erase(std::remove_if(hits_ss.begin(), hits_ss.end(),
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: why erase *and* remove_if ?

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 26]
```diff
@@ -0,0 +1,329 @@
...
+                                 [&text, &orig_query, query_length](uint64_t const hit)
+                                 {
+                                     dna4_vector matched_seq = text | ranges::view::slice(hit, hit + query_length);
+                                     return (matched_seq != orig_query);
+                                 }), hits_ss.end());
+    hits_trivial.erase(std::remove_if(hits_trivial.begin(), hits_trivial.end(),
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: please put a blank line before this

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 27]
```diff
@@ -0,0 +1,329 @@
...
+                                                      ++errors;
+                                                  else
+                                                      errors += query[i] != text[hit + i];
+                                              total_errors += errors;
+                                              if (errors != error_distribution[block])
+                                                  return true;
+                                              lb += ordered_blocks_length[block];
+                                          }
+                                          return false;
+                                      }), hits_trivial.end());
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: you put the two-line lambas outside, but define this 25-line lambda inline? :woman_facepalming:

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 28]
```diff
@@ -0,0 +1,329 @@
...
+TEST(search_scheme_test, search_scheme_edit)
+{
+    time_t seed = std::time(nullptr);
+    std::srand(seed);
+
+    test_search_scheme_edit(detail::optimum_search_scheme<0, 0>, seed, ITERATIONS);
+    test_search_scheme_edit(detail::optimum_search_scheme<0, 1>, seed, ITERATIONS);
+    test_search_scheme_edit(detail::optimum_search_scheme<0, 2>, seed, ITERATIONS);
+    test_search_scheme_edit(detail::optimum_search_scheme<0, 3>, seed, ITERATIONS);
+}
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: please undefine the macro you defined before at end of file

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+#include <seqan3/search/algorithm/detail/search_scheme_algorithm.hpp>
+#include <seqan3/search/algorithm/detail/search_trivial.hpp>
+#include <seqan3/search/fm_index/all.hpp>
+
+#include <range/v3/view/slice.hpp>
+
+#include <gtest/gtest.h>
+
+#ifdef NDEBUG
+#define ITERATIONS 1000
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: always prefix macros with `SEQAN3`, e.g `SEQAN3_SEARCH_TEST_ITERATIONS`

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 30]
```diff
@@ -0,0 +1,329 @@
...
+        bi_fm_index index(text);
+
+        uint8_t const substitution = std::rand() % (max_error + 1);
+        uint8_t const insertion    = std::rand() % (max_error + 1);
+        uint8_t const deletion     = std::rand() % (max_error + 1);
+        detail::search_param error_left{max_error, substitution, insertion, deletion};
+
+        for (uint64_t i = 0; i < iterations; ++i)
+        {
+            for (uint64_t query_length = query_length_min; query_length < query_length_max; ++query_length)
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: You are using `uint64_t` as counter variable types everywhere. Is this related to the structure of the index or could it be `size_t`, as well?

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 31]
```diff
@@ -0,0 +1,509 @@
...
+ *        backtracking.
+ * \param[in] min_error Minimum number of errors allowed.
+ * \param[in] max_error Maximum number of errors allowed.
+ */
+inline std::vector<search_dyn> compute_ss(uint8_t const min_error, uint8_t const max_error)
+{
+    // TODO: Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: We will soon have more advanced schemes computed at run-time. I think it's easier to track if we move it into a separate function from the very beginning instead of hard-coding the preliminary algorithm into the search-algorithm.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 32]
```diff
@@ -0,0 +1,509 @@
...
+ * \param[in] rb Right bound of the infix of `query` already searched (exlusive).
+ * \param[in] errors_spent Number of errors spent while searching the infix of `query`.
+ * \param[in] block_id Id of the block that infix is extended to next.
+ * \param[in] go_right The infix will be extended to the right if the flag is set to true.
+ * \param[in] min_error_left_in_block Number of remaining errors that need to be spent in the current block.
+ * \param[in] search Search of a search scheme to be used for searching.
+ * \param[in] blocks_length Cumulative block lengths of the search.
+ * \param[in] error_left Number of errors left for matching the remaining suffix of the query sequence.
+ * \param[in] delegate Function that is called on every hit. Takes `index::iterator_type` as argument.
+ */
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: There is an additional parameter.

```
* \param[in] min_error_left_in_block Number of remaining errors that need to be spent in the current block.
```

File: test/unit/search/search_scheme_algorithm_test.cpp (hunk 25 above)
Comment (cpockrandt) [reply test/unit/search/search_scheme_algorithm_test.cpp]: Remove_if doesn't remove the elements, it just moves them to the back of the range and returns the iterator of the first element to be removed.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 33]
```diff
@@ -0,0 +1,509 @@
...
+                                        error_left3, delegate);
+            }
+        } while ((go_right && it.cycle_back()) || (!go_right && it.cycle_front()));
+    }
+    return false;
+}
+
+/*!\brief Searches a query sequence in a bidirectional index using a single search of a search schemes.
+ *        Kianfar, K., Pockrandt, C., Torkamandi, B., Luo, H., & Reinert, K. (2018).
+ *        Optimum Search Schemes for Approximate String Matching Using Bidirectional FM-Index. bioRxiv, 301085.
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: I agree, it should go into details but I would prefer to keep it in this file and not move it to the top-level search function.

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 34]
```diff
@@ -0,0 +1,509 @@
...
+        if (search_ss_children<abort_on_hit>(it, query, lb, rb, errors_spent, block_id, go_right, min_error_left_in_block,
+                                             search, blocks_length, error_left, delegate) && abort_on_hit)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+/*!\brief Searches a query sequence in a bidirectional index using search schemes.
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: Search scheme is the generic term. Optimum search schemes refer only to the precomputed search schemes that are guaranteed to be optimal.

File: include/seqan3/search/algorithm/detail/search_trivial.hpp [hunk 35]
```diff
@@ -169,13 +147,25 @@ bool search_trivial(auto it, auto & query, uint64_t const query_pos, search_para
...
+* \param[in] error_left Number of errors left for matching the remaining suffix of the query sequence.
+* \param[in] delegate Function that is called on every hit. Takes `index::iterator_type` as argument.
+*
+* ### Complexity
+*
+* Exponential in the numbers of errors.
+*
+* ### Exceptions
+*
+* No-throw guarantee.
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_trivial.hpp]: I don't see any. Everything is put onto the stack.

The code in the delegate might throw an exception but this is the responsibility of the caller of `search_trivial`.

File: include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+        return pi.size();
+    }
+};
+
+//!\brief Type for storing search schemes. Number of blocks have to be known at compile time.
+template <uint8_t nbr_searches, uint8_t nbr_blocks>
+using search_scheme_type = std::array<search<nbr_blocks>, nbr_searches>;
+
+//!\brief Type for storing search schemes. Number of blocks do not have to be known at compile time.
+using search_scheme_dyn_type = std::vector<search_dyn>;
```
[the remark this answers is not in the document]
Comment (cpockrandt) [reply include/seqan3/search/algorithm/detail/search_scheme_precomputed.hpp]: This has nothing to do with optimum or non-optimum. As described in the documentation, one is used if the dimensions are known at compile time, the other if not.

File: include/seqan3/search/algorithm/search.hpp (hunk 18 above)
Comment (cpockrandt) [reply include/seqan3/search/algorithm/search.hpp]: I added a basic exception guarantee. Since queries might not be const (if it might be a range with a state), we cannot provide a strong exception guarantee.

File: include/seqan3/search/algorithm/detail/search_trivial.hpp (hunk 35 above)
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/search/algorithm/detail/search_trivial.hpp]: Ok, then please write `No-throw guarantee if this is also provided guaranteed when invoking the delegate.`

File: include/seqan3/search/algorithm/search.hpp (hunk 18 above)
Comment (h-2) [reply include/seqan3/search/algorithm/search.hpp]: You can write exactly this, e.g. "Strong exception guarantee if iterating the query does not change its state; basic exception guarantee otherwise."

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 37]
```diff
@@ -0,0 +1,519 @@
...
+    using blocks_length_type = typename search_scheme_t::value_type::blocks_length_type;
+
+    // Either store information in an array (for search schemes known at compile time) or in a vector otherwise.
+    using result_type = std::conditional_t<std::Same<search_scheme_t, search_scheme_dyn_type>,
+                                           std::vector<std::tuple<blocks_length_type, uint64_t>>,
+                                           std::array<std::tuple<blocks_length_type, uint64_t>,
+                                                      transformation_trait_or_t<std::tuple_size<search_scheme_t>,
+                                                                                std::false_type>::value>>;
+
+    bool constexpr is_dyn_scheme = std::Same<result_type, std::vector<std::tuple<blocks_length_type, uint64_t>>>;
```
Comment (h-2) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: why not `std::Same<search_scheme_t, search_scheme_dyn_type>` and before defining the result_type so that you can use it there, too? Seems more logical...

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 38]
```diff
@@ -0,0 +1,509 @@
...
+    // TODO: Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
+
+/*!\brief Returns for each search the cumulative length of blocks in the order of blocks in each search and the
+ *        starting position of the first block in the query sequence.
+ * \param[in] search_scheme Search scheme that will be used for searching.
+ * \param[in] query_length Length of the query that will be searched in an index.
+ */
```
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: This is still open :point_up:

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp (hunk 32 above)
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: You can still copy the doc and add the parameter. Or copy the doc of individual parameters, I think, maybe ask @marehr !

File: include/seqan3/search/algorithm/search.hpp [hunk 39]
```diff
@@ -95,18 +108,31 @@ inline auto search(index_t const & index, queries_t && queries, config_t const &
...
+ *
+ * ### Exceptions
+ *
+ * Basic exception guarantee.
+ */
+template <typename queries_t>
 //!\cond
     requires std::ranges::RandomAccessRange<queries_t> ||
-             (std::ranges::ForwardRange<queries_t> && std::ranges::RandomAccessRange<value_type_t<queries_t>>)
+             (std::ranges::SizedRange<queries_t> && std::ranges::RandomAccessRange<value_type_t<queries_t>>)
```
Comment (h-2) [inline include/seqan3/search/algorithm/search.hpp]: SizedRange does not imply ForwardRange so you cannot jut replace this. Please find out what your actual requirements are and specifiy those here. You need SizedRange only if you ever call size on the range. You need to random-access only if you call `[]`. If you need both, you need to require both :wink: If you just do a range-based for-loop a ForwardRange is usually sufficient.

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 40]
```diff
@@ -0,0 +1,339 @@
...
+
+    std::vector<uint64_t> hits_trivial, hits_ss;
+
+    auto delegate_trivial = [&hits_trivial](auto const & it)
+    {
+        auto const & hits_tmp = it.locate();
+        hits_trivial.insert(hits_trivial.end(), hits_tmp.begin(), hits_tmp.end());
+    };
+
+    auto delegate_ss = [&hits_ss](auto const & it)
```
Comment (h-2) [inline test/unit/search/search_scheme_algorithm_test.cpp]: we have a space between lambda capture and parameters, this affects a lot of places, but can easily be fixed...

**Comment by h-2** [comment]
Hm, somehow github got confused with my comments to earlier things and new review, I hope all comments are still there :think

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp (hunk 34 above)
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: You did not answer this question

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 41]
```diff
@@ -0,0 +1,600 @@
...
+        }
+    }
+    return false;
+}
+
+/*!\brief Searches a query sequence in a bidirectional index using search schemes.
+ *
+ * \details Reference:
+ *          Kianfar, K., Pockrandt, C., Torkamandi, B., Luo, H., & Reinert, K. (2018).
+ *          Optimum Search Schemes for Approximate String Matching Using Bidirectional FM-Index. bioRxiv, 301085.
```
Comment (h-2) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: Don't reference this everywhere, only reference it once. Also don't put details before the tparam/param documentation.

**Comment by h-2** [comment]
I am merging this now, although there are open change requests. @cpockrandt Please have a look at these and fix them when you are back.

**Comment by rrahn** [comment]
@h-2 Can you open a follow up card. It just needs to reference this merged PR.
